xdsl/transforms/stencil_to_csl_stencil.py

```diff
@@ -171,10 +295,314 @@ def match_and_rewrite(self, op: dmp.SwapOp, rewriter: PatternRewriter, /):
+    b: Sequence[Operation] = []
+    rem: Sequence[Operation] = []
+    for op in ops:
+        if isinstance(op, csl_stencil.AccessOp):
+            (b, a)[op.op == prefetch].append(op)
+        else:
+            rem.append(op)
+
+    has_changes = True
+    while (
+        len(
+            a_exports := set(
+                op
+                for op in a
+                for r in op.results
+                for u in r.uses
+                if u.operation not in a
+            )
+        )
+        > 1
+        and has_changes
+    ):
+        a_deps = set(
+            u.operation for op in a_exports for r in op.results for u in r.uses
+        )
+        has_changes = False
+        movable = [
+            op
+            for op in a_deps
+            if all(x.op in a for x in op.operands if isinstance(x, OpResult))
+        ]
+        if len(movable) > 0:
+            has_changes = True
+            for op in movable:
+                a.append(op)
+                rem.remove(op)
+
+    if len(a_exports) == 1:
+        return a, b + rem
+
+    # fallback
+    # always place `stencil.return` in second block?
+    return ops[:-1], [ops[-1]]
+
+
+@dataclass(frozen=True)
+class ConvertApplyOpPattern(RewritePattern):
+    """
+    Fuses a `csl_stencil.prefetch` and a `stencil.apply` to build a `csl_stencil.apply`.
+
+    If there are several candidate prefetch ops, the one with the largest result buffer size is selected.
+    The selection is greedy, and could in the future be expanded into a more global selection optimising for minimal
+    prefetch overhead across multiple apply ops.
+    """
+
+    num_chunks: int
+
+    def _translate_operands(
+        self, ops: Sequence[Operand], table: dict[Operand, Operand]
+    ) -> Sequence[Operand]:
+        result: list[Operand] = []
+        for operand in ops:
+            if operand in table:
+                result.append(table[operand])
+            else:
+                result.append(operand)
+        return result
+
+    def _get_insert_slice_op(
+        self,
+        source: Operand,
+        dest: Operand,
+        offset: Operand,
+        static_size: int,
+        result_type: Attribute,
+    ) -> tensor.InsertSliceOp:
+        return tensor.InsertSliceOp(
+            operands=[
+                source,
+                dest,
+                [offset],
+                [],
+                [],
+            ],
+            properties={
+                "static_offsets": DenseArrayBase.from_list(IntegerType(64), (0,)),
+                "static_sizes": DenseArrayBase.from_list(
+                    IntegerType(64),
+                    (static_size,),
+                ),
+                "static_strides": DenseArrayBase.from_list(IntegerType(64), (1,)),
+                "operandSegmentSizes": DenseArrayBase.from_list(
+                    IntegerType(32), (1, 1, 1, 0, 0)
+                ),
+            },
+            result_types=[result_type],
+        )
```

Comment:
Yes, good call. Bear in mind that this will be a static method, in order to prevent re-factoring calls in the rest of the code base that uses the existing inherited `init`, as this op is already established.